File: 600/664 - Strange Printer.java
class Solution {
    public int strangePrinter(String s) {
        int len = s.length();
        
        if (len == 0) {
            return 0;
        }
        
        int[][] dp = new int[len][len];
        
        for (int i = len - 1; i >= 0; --i) {
            for (int j = i; j < len; ++j) {
                dp[i][j] = i == j ? 1 : (1 + dp[i + 1][j]);
                
                for (int k = i + 1; k <= j ; ++k) {
                    if (s.charAt(k) == s.charAt(i)) {
                        dp[i][j] = Math.min(dp[i][j], dp[i + 1][k - 1] + dp[k][j]);
                    }
                }
            }
        }
        
        return dp[0][len - 1];
    }
}

File: csharp/664.strange-printer.cs
/*
 * @lc app=leetcode id=664 lang=csharp
 *
 * [664] Strange Printer
 */

// @lc code=start
public class Solution {
    public int StrangePrinter(string s) {
        int n = s.Length;
        if (n == 0) {
            return 0;
        }
        int[,] dp = new int[n, n];
        for (int i = n - 1; i >= 0; --i) {
            char c = s[i];
            dp[i, i] = 1;
            for (int j = i + 1; j < n; ++j) {
                dp[i, j] = 1 + dp[i + 1, j];
                for (int k = i + 1; k <= j; ++k) {
                    if (s[k] == c) {
                        dp[i, j] = Math.Min(dp[i, j], dp[k, j] + dp[i + 1, k - 1]);
                    }
                }
            }
        }
        return dp[0, n - 1];
    }
}
// @lc code=end

